### ISSUE TITLE
clang re-enables explicitly disabled diagnostics which are more specific members of a group when that group's diagnostics are enabled

### ISSUE THREAD

**Issue by nickdesaulniers** [llvm/llvm-project#63315]
Consider `-Wunused`, which itself is a group containing many sub-warnings, such as `-Wunused-but-set-variable`, and the following example code:
```c++
void foo (void) {
    int y;
    y = 0;
}
```
If you compile this with ` -Wno-unused-but-set-variable -Wunused` GCC does not warn, but clang does.  In fact, [GCC documents this behavior](https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html):

> Some options, such as -Wall and -Wextra, turn on other options, such as -Wunused, which may turn on further options, such as -Wunused-value. The combined effect of positive and negative forms is that more specific options have priority over less specific ones, independently of their position in the command-line.

The end effect of this difference is that [it makes disabling warnings portably across toolchains more difficult](https://lore.kernel.org/llvm/5fd58e1e-9e5f-4bd7-2153-aa0055a3f6c5@csgroup.eu/).

IIRC, clang re-enables all of the subgroups of `-Wunused`, without first checking whether any of the subgroups were previously explicitly disabled.  Is this a known issue (I wasn't able to find a pre-existing bug report)?

**Comment by AaronBallman** [comment]
Oof. IIRC, our policy is "last flag wins" (though I need to double-check that). But I'm mostly worried that compatibility with GCC would then change behavior for our users in ways that aren't acceptable. e.g.,

`clang -Wno-unused-but-set-variable -Wunused foo.c` might be giving the expected behavior for the user because the default build system may be setting `-Wno-unused-but-set-variable` while the user adds a CFLAGS override to add `-Wunused` to enable the warning.

My instinct is that we need to proceed with caution.

**Comment by nickdesaulniers** [comment]
@tstellar found https://discourse.llvm.org/t/warnings-difference-between-gcc-and-clang/38515.
